### PR TITLE
feat: Add an App Menu to the Header Component

### PR DESCRIPTION
**Pull request by Carlos-Muniz**
StudioHeader is a graft of the Header Component with an additional optional AppMenu.
Some Frontend Apps use Menus in their custom headers to provide more
functionality in their apps. By adding this functionality in
StudioHeader, it will be easier for frontend apps in Studio to adopt
this component without it affecting the main Header component.

### CODE REVIEW

**Review by arbrandes — CHANGES_REQUESTED** [review]
@Carlos-Muniz, I'm impressed.  This is great work.  I'm having a hard time finding stuff to nitpick!  Thanks for the tests, too.

That said, I tried this out in frontend-app-library-authoring, and it doesn't solve our problem completely.  I think we're going to need a `<StudioHeader />` component after all.  Let's get together tomorrow and discuss it.

Worry not, though: as best I can tell, the work you've done here so far stays.  We'll just have to add a few things.

File: src/StudioHeader.jsx [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+      href: `${config.STUDIO_BASE_URL}/dashboard`,
+      content: intl.formatMessage(messages['header.links.courses']),
+    },
+  ];
+
+  const orderHistoryItem = {
+    type: 'item',
+    href: config.ORDER_HISTORY_URL,
+    content: intl.formatMessage(messages['header.user.menu.order.history']),
+  };
```
Comment (arbrandes) [inline src/StudioHeader.jsx:48]: This can go.  Orders are of no concern to Studio users.

File: src/StudioHeader.jsx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+  return (
+    <>
+      <Responsive maxWidth={768}>
+        <DesktopHeader {...props} />
+      </Responsive>
+      <Responsive minWidth={769}>
+        <DesktopHeader {...props} />
+      </Responsive>
```
Comment (arbrandes) [inline src/StudioHeader.jsx:111]: Studio doesn't really do mobile, so we can ditch the ` <Responsive />` stuff here.

File: src/StudioHeader.jsx [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      type: 'item',
+      href: config.LOGOUT_URL,
+      content: intl.formatMessage(messages['header.user.menu.logout']),
+    },
+  ];
+
+  // Users should only see Order History if have a ORDER_HISTORY_URL define in the environment.
+  if (config.ORDER_HISTORY_URL) {
+    userMenu.splice(-1, 0, orderHistoryItem);
+  }
```
Comment (arbrandes) [inline src/StudioHeader.jsx:76]: This can also go.

File: src/StudioHeader.jsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  const orderHistoryItem = {
+    type: 'item',
+    href: config.ORDER_HISTORY_URL,
+    content: intl.formatMessage(messages['header.user.menu.order.history']),
+  };
+
+  const userMenu = authenticatedUser === null ? [] : [
+    {
+      type: 'item',
+      href: `${config.STUDIO_BASE_URL}/dashboard`,
```
Comment (arbrandes) [inline src/StudioHeader.jsx:53]: As above, `/dashboard` should be removed.

File: src/StudioHeader.jsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+      content: intl.formatMessage(messages['header.user.menu.dashboard']),
```
Comment (arbrandes) [inline src/StudioHeader.jsx:54]: We're going to need a new message, here.  It should say "Studio Home", not "Dashboard".

File: src/StudioHeader.jsx [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    {
+      type: 'item',
+      href: `${config.STUDIO_BASE_URL}/u/${authenticatedUser.username}`,
+      content: intl.formatMessage(messages['header.user.menu.profile']),
+    },
+    {
+      type: 'item',
+      href: `${config.STUDIO_BASE_URL}/account/settings`,
+      content: intl.formatMessage(messages['header.user.menu.account.settings']),
+    },
```
Comment (arbrandes) [inline src/StudioHeader.jsx:65]: We can ditch these two menu items.  In their stead, we should have a link to the maintenance page at `${config.STUDIO_BASE_URL}/maintenance`.   See the equivalent implementation here:

https://github.com/openedx/frontend-app-library-authoring/blob/master/src/library-authoring/studio-header/StudioHeader.jsx#L87

File: src/StudioHeader.jsx [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      type: 'item',
+      href: `${config.STUDIO_BASE_URL}/register`,
+      content: intl.formatMessage(messages['header.user.menu.register']),
+    },
+  ];
+
+  const props = {
+    logo: config.LOGO_URL,
+    logoAltText: config.SITE_NAME,
+    logoDestination: `${config.STUDIO_BASE_URL}/dashboard`,
```
Comment (arbrandes) [inline src/StudioHeader.jsx:94]: Another instance of `/dashboard` that has to be removed.

File: src/StudioHeader.jsx [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+      type: 'item',
+      href: config.LOGIN_URL,
+      content: intl.formatMessage(messages['header.user.menu.login']),
+    },
+    {
+      type: 'item',
+      href: `${config.STUDIO_BASE_URL}/register`,
+      content: intl.formatMessage(messages['header.user.menu.register']),
+    },
+  ];
```
Comment (arbrandes) [inline src/StudioHeader.jsx:89]: No Studio MFE (like library-authoring) will render anything when being accessed anonymously (they redirect to the login page immediately), so we can ditch `loggedOutItems`.

File: src/StudioHeader.jsx [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+  const props = {
+    logo: config.LOGO_URL,
+    logoAltText: config.SITE_NAME,
+    logoDestination: `${config.STUDIO_BASE_URL}/dashboard`,
+    loggedIn: authenticatedUser !== null,
+    username: authenticatedUser !== null ? authenticatedUser.username : null,
+    avatar: authenticatedUser !== null ? authenticatedUser.avatar : null,
+    mainMenu: getConfig().AUTHN_MINIMAL_HEADER ? [] : mainMenu,
+    userMenu: getConfig().AUTHN_MINIMAL_HEADER ? [] : userMenu,
```
Comment (arbrandes) [inline src/StudioHeader.jsx:99]: I don't think we need the `AUTHN_MINIMAL_HEADER` stuff for Studio MFEs.

File: src/StudioHeader.test.jsx [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+            <StudioHeader appMenu={appMenu} />
+          </AppContext.Provider>
+        </IntlProvider>
+      </ResponsiveContext.Provider>
+    );
+
+    const wrapper = TestRenderer.create(component);
+
+    expect(wrapper.toJSON()).toMatchSnapshot();
+  });
```
Comment (arbrandes) [inline src/StudioHeader.test.jsx:120]: As noted above, don't think there's a need to test anonymous access, here.  It's not likely to ever happen.

File: src/StudioHeader.test.jsx [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+            <StudioHeader />
+          </AppContext.Provider>
+        </IntlProvider>
+      </ResponsiveContext.Provider>
+    );
+
+    const wrapper = TestRenderer.create(component);
+
+    expect(wrapper.toJSON()).toMatchSnapshot();
+  });
```
Comment (arbrandes) [inline src/StudioHeader.test.jsx:35]: No need for anonymous testing.

File: src/StudioHeader.test.jsx [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+  it('renders correctly for authenticated desktop', () => {
+    const component = (
+      <ResponsiveContext.Provider value={{ width: 1280 }}>
```
Comment (arbrandes) [inline src/StudioHeader.test.jsx:39]: No need for the responsiveness context: Studio-stuff (including the library authoring MFE) isn't mobile-ready.

File: src/StudioHeader.jsx [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+function StudioHeader({ intl, appMenu }) {
+  const { authenticatedUser, config } = useContext(AppContext);
+
+  const mainMenu = [
+    {
+      type: 'item',
+      href: `${config.STUDIO_BASE_URL}/dashboard`,
+      content: intl.formatMessage(messages['header.links.courses']),
+    },
+  ];
```
Comment (arbrandes) [inline src/StudioHeader.jsx:42]: `mainMenu` should be an optional prop (empty by default).  In frontend-app-library-authoring, for instance, we'll need to populate it with a link to the selected library.  See here:

https://github.com/openedx/frontend-app-library-authoring/blob/master/src/library-authoring/studio-header/StudioHeader.jsx#L39-L43

**Review by arbrandes — CHANGES_REQUESTED** [review]
The structure looks good, thanks!  We'll need to make some adjustments so it works like the current library-authoring header (which in turn mimics the actual Studio header), though.  See the inline comments.

File: src/StudioHeader.jsx (hunk 1 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.jsx]: resolved

File: src/StudioHeader.jsx (hunk 2 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.jsx]: removed responsive components

File: src/StudioHeader.jsx (hunk 3 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.jsx]: removed

File: src/StudioHeader.jsx (hunk 4 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.jsx]: removed

File: src/StudioHeader.jsx (hunk 5 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.jsx]: wrote new message

File: src/StudioHeader.jsx (hunk 7 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.jsx]: removed

File: src/StudioHeader.jsx (hunk 8 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.jsx]: ditched

File: src/StudioHeader.jsx (hunk 9 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.jsx]: removed

File: src/StudioHeader.test.jsx (hunk 10 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.test.jsx]: tests removed

File: src/StudioHeader.jsx (hunk 13 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.jsx]: mainmenu is now an optional prop

File: src/StudioHeader.test.jsx (hunk 11 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.test.jsx]: anonymous testing removed

File: src/StudioHeader.test.jsx (hunk 12 above)
Comment (Carlos-Muniz) [reply src/StudioHeader.test.jsx]: Removed

File: src/DesktopHeader.jsx [hunk 14]
```diff
@@ -54,6 +54,24 @@ class DesktopHeader extends React.Component {
     });
   }
 
+  // Renders an optional App Menu for
+  renderAppMenu() {
+    const { appMenu } = this.props;
+    const { content, href, menuItems } = appMenu;
+    return (
+      <Menu transitionClassName="menu-dropdown" transitionTimeout={250}>
+        <MenuTrigger tag="a" className="nav-link d-inline-flex align-items-center" href={href}>
```
Comment (arbrandes) [inline src/DesktopHeader.jsx:63]: I think we can drop the `href`.  Having a link that also works as a dropdown can be midly confusing.

**Review by arbrandes — CHANGES_REQUESTED** [review]
Almost there!  Just a couple of things, and I think we'll be good to go.

File: src/DesktopHeader.jsx (hunk 14 above)
Comment (Carlos-Muniz) [reply src/DesktopHeader.jsx]: href removed!

File: src/DesktopHeader.jsx [hunk 15]
```diff
@@ -54,6 +54,24 @@ class DesktopHeader extends React.Component {
...
+          {content} <CaretIcon role="img" aria-hidden focusable="false" />
+        </MenuTrigger>
+        <MenuContent className="mb-0 dropdown-menu show dropdown-menu-right pin-right shadow py-2">
+          {menuItems.map((menuType, menuHref, menuContent) => (
```
[the remark this answers is not in the document]
Comment (Carlos-Muniz) [reply src/DesktopHeader.jsx]: Added the {} around the triplet and renamed them to stick with the pattern.

**Review by arbrandes — APPROVED** [review]
Thanks for bearing with me, @Carlos-Muniz!  LGTM!

**Pull request merged**